### PR TITLE
Make repository cache settings configurable using repository server CR

### PR DESCRIPTION
**Pull request by kale-amruta**
## Change Overview

Currently the metadata/content cache settings required while connecting to kopia repository are not configurable in the repository server controller. This PR makes it configurable through CR

## Pull request type

Please check the type of change your PR introduces:
- [ ] :construction: Work in Progress
- [ ] :rainbow: Refactoring (no functional changes, no api changes)
- [ ] :hamster: Trivial/Minor
- [ ] :bug: Bugfix
- [ ] :sunflower: Feature
- [ ] :world_map: Documentation
- [ ] :robot: Test

## Issues 

- fixes #issue-number

## Test Plan

- [ ] :muscle: Manual
- [ ] :zap: Unit test
- [ ] :green_heart: E2E

### CODE REVIEW

File: pkg/customresource/repositoryserver.yaml [hunk 1]
```diff
@@ -36,6 +36,18 @@ spec:
                 description: Repository details for the purpose of establishing a
                   connection
                 properties:
+                  cacheSizeSettings:
+                    description: CacheSettings are the metadata/content cache size
+                      details that can be used while initializing kopia repository
```
Comment (ankitjain235) [inline pkg/customresource/repositoryserver.yaml:41]:
```suggestion
                    description: CacheSettings are the metadata/content cache size
                      details that can be used while establishing connection to the kopia repository
```

File: go.mod [hunk 2]
```diff
@@ -63,7 +63,13 @@ require (
...
 )
 
-require github.com/dustin/go-humanize v1.0.1
+require (
+	github.com/dustin/go-humanize v1.0.1
+	github.com/go-logr/logr v1.2.0
+	github.com/onsi/ginkgo/v2 v2.0.0
+	github.com/onsi/gomega v1.18.1
+	sigs.k8s.io/kustomize/kyaml v0.13.6
+)
```
Comment (ankitjain235) [inline go.mod:72]: Why are these changes here? I cannot find any new imports in packages.

**Review by ankitjain235 — COMMENTED** [review]
LGTM overall, added minor comments.

File: pkg/customresource/repositoryserver.yaml [hunk 3]
```diff
@@ -37,6 +37,18 @@ spec:
                 description: Repository has the details required by the repository
                   server to connect to kopia repository
                 properties:
+                  cacheSizeSettings:
+                    description: CacheSettings are the metadata/content cache size
+                      details that can be used establishing connection to the kopia repository
```
Comment (ankitjain235) [inline pkg/customresource/repositoryserver.yaml:42]:
```suggestion
                      details that can be used for establishing connection to the kopia repository
```

**Review by ankitjain235 — APPROVED** [review]
LGTM besides minor comment

File: pkg/controllers/repositoryserver/repository.go [hunk 4]
```diff
@@ -52,3 +57,20 @@ func (h *RepoServerHandler) connectToKopiaRepository() error {
...
+func (h *RepoServerHandler) getRepositoryCacheSettings() (contentCacheMB, metadataCacheMB int, err error) {
+	contentCacheMB, metadataCacheMB = command.GetGeneralCacheSizeSettings()
+	if h.RepositoryServer.Spec.Repository.CacheSizeSettings.Content != "" {
+		contentCacheMB, err = strconv.Atoi(h.RepositoryServer.Spec.Repository.CacheSizeSettings.Content)
+		if err != nil {
+			return
+		}
+	}
+	if h.RepositoryServer.Spec.Repository.CacheSizeSettings.Metadata != "" {
+		contentCacheMB, err = strconv.Atoi(h.RepositoryServer.Spec.Repository.CacheSizeSettings.Metadata)
```
Comment (pavannd1) [inline pkg/controllers/repositoryserver/repository.go:70]:
```suggestion
		metadataCacheMB, err = strconv.Atoi(h.RepositoryServer.Spec.Repository.CacheSizeSettings.Metadata)
```

**Review by pavannd1 — CHANGES_REQUESTED** [review]

File: pkg/controllers/repositoryserver/repository.go [hunk 5]
```diff
@@ -52,3 +57,20 @@ func (h *RepoServerHandler) connectToKopiaRepository() error {
 		args,
 	)
 }
+
+func (h *RepoServerHandler) getRepositoryCacheSettings() (contentCacheMB, metadataCacheMB int, err error) {
```
Comment (pavannd1) [inline pkg/controllers/repositoryserver/repository.go:61]: Reminder to add a unit test for this along with other tests for repo server controller

File: pkg/controllers/repositoryserver/repository.go [hunk 6]
```diff
@@ -52,3 +57,20 @@ func (h *RepoServerHandler) connectToKopiaRepository() error {
...
+	contentCacheMB, metadataCacheMB = command.GetGeneralCacheSizeSettings()
+	if h.RepositoryServer.Spec.Repository.CacheSizeSettings.Content != "" {
```
Comment (ankitjain235) [inline pkg/controllers/repositoryserver/repository.go:63]: Can `CacheSizeSettings` also be nil here and needs a nil check?

File: pkg/apis/cr/v1alpha1/repositoryserver_types.go [hunk 7]
```diff
@@ -76,6 +76,14 @@ type Repository struct {
 	Hostname string `json:"hostname"`
 	// PasswordSecretRef has the password required to connect to kopia repository
 	PasswordSecretRef corev1.SecretReference `json:"passwordSecretRef"`
+	CacheSizeSettings CacheSizeSettings      `json:"cacheSizeSettings"`
```
Comment (ankitjain235) [inline pkg/apis/cr/v1alpha1/repositoryserver_types.go:79]: Should we set omitEmpty here or mark it as required in CRD?

**Review by ankitjain235 — COMMENTED** [review]
Minor questions, looks good otherwise.

File: pkg/apis/cr/v1alpha1/repositoryserver_types.go (hunk 7 above)
Comment (kale-amruta) [reply pkg/apis/cr/v1alpha1/repositoryserver_types.go]: added it

File: pkg/controllers/repositoryserver/repository.go (hunk 6 above)
Comment (kale-amruta) [reply pkg/controllers/repositoryserver/repository.go:63]: its not a pointer, hence it cannot be nil. it will always have `{}` value when its unmarshalled even if its not present in the CRD.

**Review by ankitjain235 — APPROVED** [review]
LGTM

File: pkg/controllers/repositoryserver/repository.go (hunk 6 above)
Comment (pavannd1) [inline pkg/controllers/repositoryserver/repository.go:63]: What happens here when the CR does not have `CacheSizeSettings`? Would it resolve to `{Content: "", Metadata: ""}`?

Comment (kale-amruta) [reply pkg/controllers/repositoryserver/repository.go:63]: yes @pavannd1

**Review by pavannd1 — APPROVED** [review]

**Comment by PrasadG193** [comment]
@kale-amruta we need to run codegen after these changes. Can you please raise a PR?
